src/views/sources/addSourceModal.tsx

```diff
@@ -31,34 +31,69 @@ interface AddSourceModalProps {
   source?: SourceType;
   sourceType?: string;
   onClose?: () => void;
-  onSubmit?: (payload) => void;
-  useGetCredentials?: typeof useGetCredentialsApi;
+  onSubmit?: (payload: any) => void;
 }
 
-const AddSourceModal: React.FC<AddSourceModalProps> = ({
-  isOpen,
-  source,
+interface SourceFormProps extends Omit<AddSourceModalProps, 'isOpen'> {
+  useForm?: typeof useSourceForm;
+}
+
+interface SourceFormType {
+  credentials?: number[];
+  useParamiko?: boolean;
+  sslVerify?: boolean;
+  sslProtocol: string;
+  name?: string;
+  hosts?: string;
+  port?: string;
+}
+
+const useSourceForm = ({
   sourceType,
-  onClose = Function.prototype,
-  onSubmit = Function.prototype,
+  source,
   useGetCredentials = useGetCredentialsApi
-}) => {
+}: { sourceType?: string; source?: Partial<SourceType>; useGetCredentials?: typeof useGetCredentialsApi } = {}) => {
+  const initialFormState: SourceFormType = {
+    credentials: [],
+    useParamiko: false,
+    sslVerify: true,
+    sslProtocol: 'SSLv23',
+    name: '',
+    hosts: '',
+    port: ''
+  };
+
   const { getCredentials } = useGetCredentials();
   const [credOptions, setCredOptions] = useState<{ value: string; label: string }[] | []>([]);
-  const [credentials, setCredentials] = useState<number[]>(source?.credentials?.map(c => c.id) || []);
-  const [useParamiko, setUseParamiko] = useState<boolean>(source?.options?.use_paramiko ?? false);
-  const [sslVerify, setSslVerify] = useState<boolean>(source?.options?.ssl_cert_verify ?? true);
-  const [sslProtocol, setSslProtocol] = useState<string>(
-    source?.options?.disable_ssl ? 'Disable SSL' : source?.options?.ssl_protocol || 'SSLv23'
-  );
+  const [formData, setFormData] = useState<SourceFormType>(initialFormState);
+
+  const typeValue = source?.source_type || sourceType?.split(' ')?.shift()?.toLowerCase();
+  const isNetwork = typeValue === 'network';
 
-  const sourceTypeValue = source?.source_type || sourceType?.split(' ')?.shift()?.toLowerCase();
-  const isNetwork = sourceTypeValue === 'network';
+  // Edit props, reset state on unmount
+  useEffect(() => {
+    if (source) {
+      setFormData({
+        credentials: source?.credentials?.map(c => c.id) || [],
```

Comment:
Do we need that `source?` everywhere? This entire block is inside `if (source)`, so I would assume we can count on `source` being defined.

Reply:
we actually do, we use this for both adding and editing sources. we set the fields we would like to be returned, but if we are creating a new source we will set the fields with empy values : setFormData(initialFormState);